Marlin/src/HAL/HAL_DUE/fastio_Due.h

```diff
@@ -0,0 +1,436 @@
+*/
+
+/**
+ * Description: Fast IO functions for Arduino Due and compatible (SAM3X8E)
+ *
+ * For ARDUINO_ARCH_SAM
+ */
+
+#ifndef	_FASTIO_DUE_H
+#define	_FASTIO_DUE_H
+
+/**
+  utility functions
+*/
+
+#ifndef MASK
+  #define MASK(PIN)  (1 << PIN)
+#endif
+
+/**
+  magic I/O routines
+  now you can simply SET_OUTPUT(STEP); WRITE(STEP, 1); WRITE(STEP, 0);
+*/
+
+/// Read a pin
+#define _READ(IO) ((bool)(DIO ## IO ## _WPORT -> PIO_PDSR & (MASK(DIO ## IO ## _PIN))))
+
+/// Write to a pin
+#define _WRITE_VAR(IO, v) do {  if (v) {g_APinDescription[IO].pPort->PIO_SODR = g_APinDescription[IO].ulPin; } \
+                                    else {g_APinDescription[IO].pPort->PIO_CODR = g_APinDescription[IO].ulPin; } \
+                                 } while (0)
+
+#define _WRITE(IO, v) do {  if (v) {DIO ## IO ## _WPORT -> PIO_SODR = MASK(DIO ## IO ##_PIN); } \
+                                else {DIO ##  IO ## _WPORT -> PIO_CODR = MASK(DIO ## IO ## _PIN); }; \
+                             } while (0)
+
+/// toggle a pin
+#define _TOGGLE(IO)  _WRITE(IO, !READ(IO))
+
+/// set pin as input
+#define _SET_INPUT(IO)  pmc_enable_periph_clk(g_APinDescription[IO].ulPeripheralId); \
+                        PIO_Configure(g_APinDescription[IO].pPort, PIO_INPUT, g_APinDescription[IO].ulPin, 0)
+/// set pin as output
+#define _SET_OUTPUT(IO)  PIO_Configure(g_APinDescription[IO].pPort, PIO_OUTPUT_1, \
+                         g_APinDescription[IO].ulPin, g_APinDescription[IO].ulPinConfiguration)
+
+/// set pin as input with pullup mode
+#define _PULLUP(IO, v)  { pinMode(IO, (v!=LOW ? INPUT_PULLUP : INPUT)); }
+
+/// check if pin is an input
+#define _GET_INPUT(IO)
+/// check if pin is an output
+#define _GET_OUTPUT(IO)
+
+/// check if pin is an timer
+#define _GET_TIMER(IO)
+
+//  why double up on these macros? see http://gcc.gnu.org/onlinedocs/cpp/Stringification.html
+
+/// Read a pin wrapper
+#define READ(IO)  _READ(IO)
+
+/// Write to a pin wrapper
+#define WRITE_VAR(IO, v)  _WRITE_VAR(IO, v)
+#define WRITE(IO, v)  _WRITE(IO, v)
+
+/// toggle a pin wrapper
+#define TOGGLE(IO)  _TOGGLE(IO)
+
+/// set pin as input wrapper
+#define SET_INPUT(IO)  _SET_INPUT(IO)
+/// set pin as input with pullup wrapper
+#define SET_INPUT_PULLUP(IO) do{ _SET_INPUT(IO); _PULLUP(IO, HIGH); }while(0)
+/// set pin as output wrapper
+#define SET_OUTPUT(IO)  do{ _SET_OUTPUT(IO); _WRITE(IO, LOW); }while(0)
+
+/// check if pin is an input wrapper
+#define GET_INPUT(IO)  _GET_INPUT(IO)
+/// check if pin is an output wrapper
+#define GET_OUTPUT(IO)  _GET_OUTPUT(IO)
+
+/// check if pin is an timer wrapper
+#define GET_TIMER(IO)  _GET_TIMER(IO)
+
+// Shorthand
+#define OUT_WRITE(IO, v) { SET_OUTPUT(IO); WRITE(IO, v); }
+
+/**
+  ports and functions
+
+  added as necessary or if I feel like it- not a comprehensive list!
+*/
+
+// UART
+#define RXD        DIO0
+#define TXD        DIO1
+
+// TWI (I2C)
+#define SCL        DIO21
+#define SDA        DIO20
+
+/**
+pins
+*/
+
+#define DIO0_PIN 8
```

Comment:
I just want to inform you that this was some work years ago. Don't forget to contrubute.
https://github.com/Wurstnase/Marlin4Due/commit/5e1df35cca5686 

```
Copyright (c) 2014 Bob Cousins bobcousins42@googlemail.com
Copyright (c) 2015-2016 Nico Tonnhofer wurstnase.reprap@gmail.com

```